File: deep_search-dev/api/llm_api.py
import asyncio
import os
from dotenv import load_dotenv
from openai import AsyncOpenAI
import logging


# 设置日志级别
logging.getLogger("openai").setLevel(logging.ERROR)

# 加载环境变量
load_dotenv()


class LLM():

    def __init__(self, logger=None):
        self.model = os.getenv('GLM_MODEL', 'glm-4.5-flash')
        self.client = AsyncOpenAI(
            api_key=os.getenv('GLM_API_KEY'),
            base_url=os.getenv('GLM_BASE_URL', 'https://openai.deepseek.cn/v1')
        )
        self.current_request = None
        self.logger = logger or logging.getLogger(__name__)

    async def infer(self, prompt, enable_thinking=False, temperature=0.7):
        # 确保打印的 self.model 不是元组
        self.logger.info(f'调用模型：{self.model}......')
        
        # 创建可中断的请求
        async with asyncio.TaskGroup() as tg:
            request_task = tg.create_task(self.client.chat.completions.create(
                model=self.model,
                messages=[{"role": "user", "content": prompt}],
                extra_body={
                    'thinking': {
                        "type": "enabled" if enable_thinking else "disabled"
                    }
                },
                temperature=temperature
            ))

            # 保存当前请求的引用
            self.current_request = request_task

            try:
                response = await request_task
                content = response.choices[0].message.content
                return content
            except Exception as e:
                if 'You exceeded your current quota, please check your plan and billing details.' in str(e):
                    content = 'You exceeded your current quota, please check your plan and billing details.'
                elif '413 Request Entity Too Large' in str(e):
                    content = '413 Request Entity Too Large.'
                else:
                    raise Exception(f'llm fail: Too many requests or Other error.')
                return content
            finally:
                # 清除请求引用
                self.current_request = None

    def cancel_request(self):
        """取消当前请求"""
        if self.current_request and not self.current_request.done():
            self.current_request.cancel()
            return True
        return False


llm = LLM()

File: deep_search-dev-go/api/llm_api.go
package api

import (
	"bytes"
	"context"
	"deepsearch/logs"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"sync"

	"github.com/sashabaranov/go-openai"
)

type LLM struct {
	client        *openai.Client
	model         string
	currentCancel context.CancelFunc
	logger        *logs.Logger
	mu            sync.Mutex
}

func NewLLM(logger *logs.Logger) *LLM {
	config := openai.DefaultConfig(os.Getenv("GLM_API_KEY"))
	config.BaseURL = os.Getenv("GLM_BASE_URL")

	model := os.Getenv("GLM_MODEL")
	if model == "" {
		model = "glm-4.5-flash"
	}

	return &LLM{
		client: openai.NewClientWithConfig(config),
		model:  model,
		logger: logger,
	}
}

func (llm *LLM) Infer(ctx context.Context, prompt string, enableThinking bool, temperature float32) (string, error) {
	llm.mu.Lock()
	defer llm.mu.Unlock()

	llm.logger.Infof("调用模型：%s......", llm.model)

	reqCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	llm.currentCancel = cancel

	// 统一使用自定义HTTP请求，以便控制思考模式参数
	return llm.sendRequestWithThinkingMode(reqCtx, prompt, temperature, enableThinking)
}

// sendRequestWithThinkingMode 发送带思考模式参数的请求
func (llm *LLM) sendRequestWithThinkingMode(ctx context.Context, prompt string, temperature float32, enableThinking bool) (string, error) {
	// 根据enableThinking参数设置思考模式
	thinkingType := "disabled"
	if enableThinking {
		thinkingType = "enabled"
		llm.logger.Info("启用思考模式")
	} else {
		llm.logger.Info("禁用思考模式")
	}

	// 构建请求数据
	requestData := map[string]interface{}{
		"model": llm.model,
		"messages": []map[string]string{
			{
				"role":    "user",
				"content": prompt,
			},
		},
		"thinking": map[string]string{
			"type": thinkingType,
		},
		"max_tokens":  4096,
		"temperature": temperature,
	}

	// 序列化请求数据
	jsonData, err := json.Marshal(requestData)
	if err != nil {
		llm.logger.Error(fmt.Sprintf("序列化请求数据失败: %v", err))
		return "", fmt.Errorf("failed to marshal request data: %v", err)
	}

	// 获取配置信息（通过环境变量重新创建）
	baseURL := os.Getenv("GLM_BASE_URL")
	apiKey := os.Getenv("GLM_API_KEY")

	// 创建自定义HTTP客户端
	httpClient := &http.Client{
		Timeout: 0, // 不设置超时，完全由上下文控制
	}

	// 确保URL格式正确（去除多余的斜杠）
	endpoint := fmt.Sprintf("%s/chat/completions", strings.TrimSuffix(baseURL, "/"))

	// 创建自定义HTTP请求
	httpReq, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		llm.logger.Error(fmt.Sprintf("创建HTTP请求失败: %v", err))
		return "", fmt.Errorf("failed to create HTTP request: %v", err)
	}

	// 设置请求头
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Authorization", fmt.Sprintf("Bearer %s", apiKey))

	// 发送请求
	llm.logger.Info("发送请求...")
	httpResp, err := httpClient.Do(httpReq.WithContext(ctx))
	if err != nil {
		llm.logger.Error(fmt.Sprintf("发送HTTP请求失败: %v", err))
		return "", fmt.Errorf("failed to send HTTP request: %v", err)
	}
	defer httpResp.Body.Close()

	// 读取响应体
	respBody, err := io.ReadAll(httpResp.Body)
	if err != nil {
		llm.logger.Error(fmt.Sprintf("读取响应体失败: %v", err))
		return "", fmt.Errorf("failed to read response body: %v", err)
	}

	// 检查HTTP状态码
	if httpResp.StatusCode != http.StatusOK {
		llm.logger.Error(fmt.Sprintf("API请求失败，状态码: %d, 响应: %s", httpResp.StatusCode, string(respBody)))
		return "", fmt.Errorf("API request failed with status code %d: %s", httpResp.StatusCode, string(respBody))
	}

	// 解析响应
	var respData struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}

	if err := json.Unmarshal(respBody, &respData); err != nil {
		llm.logger.Error(fmt.Sprintf("解析响应数据失败: %v", err))
		return "", fmt.Errorf("failed to unmarshal response data: %v", err)
	}

	// 清除取消函数
	llm.currentCancel = nil

	if len(respData.Choices) == 0 {
		return "", fmt.Errorf("no response from LLM")
	}

	return respData.Choices[0].Message.Content, nil
}

// CancelRequest 取消当前请求
func (llm *LLM) CancelRequest() bool {
	llm.mu.Lock()
	defer llm.mu.Unlock()

	if llm.currentCancel != nil {
		llm.currentCancel()
		llm.currentCancel = nil
		return true
	}
	return false
}

// 全局 LLM 实例
var GlobalLLM *LLM

// InitLLM 初始化全局 LLM 实例
func InitLLM(logger *logs.Logger) {
	GlobalLLM = NewLLM(logger)
}
